### PR TITLE
Blog App - processing data in models

### PR DESCRIPTION
**Pull request by khanh-devos**
### Create at least one user by running the following code:
- [x] first_user = User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')
- [x] second_user = User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Poland.'
- [x] Create at least 4 posts written by one of the users you created by running the following code:
   first_post = Post.create(author: first_user, title: 'Hello', text: 'This is my first post')
- [x] Create at least 6 posts comments for one of the posts you created by running the following code:
   Comment.create(post: first_post, author: second_user, text: 'Hi Tom!' )
- [x] Use other [CRUD methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/active_record_crud.md) to find, update and delete entities.
[Create custom methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/Processing_data_model_project_v1_1.md#create-custom-methods)

#### Users
- [x] A method that returns the 3 most recent posts for a given user.
#### Posts
- [x] A method that updates the posts counter for a user.
- [x] A method which returns the 5 most recent comments for a given post.
#### Comments
- [x]  A method that updates the comments counter for a post.
#### Likes
- [x] A method that updates the likes counter for a post.
- [x] Go to rails c and check if your methods are working.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,7 +1,11 @@
 class User < ApplicationRecord
   has_one_attached :photo
 
-  has_many :comments
-  has_many :likes
-  has_many :posts, foreign_key: :author_id
+  has_many :comments, foreign_key: 'user_id'
+  has_many :likes, foreign_key: 'user_id'
+  has_many :posts, foreign_key: 'author_id'
```
Comment (TadesseDev) [inline app/models/user.rb:6]: - [Optional] Good job using the `author_id` for the association between the `post` and `user` models, but the same structure should be preserved for the `comments` and `likes` association. Please consider following the same structure and changing the `user_id` to `authot_id`. Remember when working with a ERD diagram we are expected to create columns with the same name.

File: app/models/post.rb [hunk 2]
```diff
@@ -1,6 +1,14 @@
...
+  has_many :comments, foreign_key: 'post_id'
+  has_many :likes, foreign_key: 'post_id'
 
-  belongs_to :users
+  belongs_to :user, foreign_key: 'author_id'
+
+  def self.count_posts(user)
+    user.update(postCounter: user.posts.size)
+  end
+
```
Comment (TadesseDev) [inline app/models/post.rb]: - [Optional] Dear @khanh-devos, for such a method which is expected to run every time we create a new post it's a better practice to apply the `after_save` helper. This way we can keep our data consistent and we don't need to call the `count_posts` method to update the author post counter, This also goes for all the other methods inside the `Like` and `Comment` controllers.  [see here for more](https://guides.rubyonrails.org/active_record_callbacks.html)

File: app/models/post.rb [hunk 3]
```diff
@@ -1,6 +1,14 @@
 class Post < ApplicationRecord
-  has_many :comments
-  has_many :likes
+  has_many :comments, foreign_key: 'post_id'
+  has_many :likes, foreign_key: 'post_id'
 
-  belongs_to :users
+  belongs_to :user, foreign_key: 'author_id'
+
+  def self.count_posts(user)
```
Comment (TadesseDev) [inline app/models/post.rb]: - [Optional] Dear @khanh-devos, using `self` while creating a method makes the method a `static` method, which means the method `belongs to the class` and not to the `object/user` on which the method is called. Hence every time you call such a method you are expected to pass the user. Instead, a better implementation is to create a nonstatic method/instance method/without `self.method` and use the attribute of the object to perform your operation. 
**This is one of the most powerful weapons of OOP**  
**[Please refer here about static and nonstatic/instance members of the class](https://www.geeksforgeeks.org/ruby-static-members/)**

## Here is a preferable implementation 
```ruby
 def update_post_counter
    author.increment!(:posts_counter)  
  end
```

**Review by TadesseDev — APPROVED** [review]
### Hi @khanh-devos 

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
AmazingCreepyFrogmouth-max-1mb




## Highlights
 - [x] The associations between models are set up correctly. ✅ 
 - [x] All expected methods are in place. ✅

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 3 above)
Comment (khanh-devos) [reply app/models/post.rb]: Thank you for your review @TadesseDev

**Pull request merged**